### DISCUSSION TITLE
How do I start from scratch?

### DISCUSSION THREAD

**Discussion by IngaleAbhishek** [eclipse-emfcloud/emfcloud#110] [All other] [answered]
Hi,

I intend to develop a modeling tool for my custom usecase. I have created an ecore diagram and the corresponding genmodel, along with the generated code files (The Factory, Package, Impl, and Switch classes) using the Eclipse IDE. I intend to replace the provided example ecore file with my newly created ecore file and its corressponding classes, and start afresh.

The final aim is to be able to draw a diagram as per the usecase, while keeping the serialization format as **XML** instead of JSON.

I refered the **Workflow Diagram example** to reuse the existing stub. I could figure out that the ecore and genmodel file needs to be placed in the **model** directory, and the generated classes (classes generated in Eclipse IDE) to be placed in the **src-gen** directory.

_Server code downloaded from the following URL_: https://github.com/eclipse-glsp/glsp-server
_Client code downloaded from the following URL_  : https://github.com/eclipse-glsp/glsp-theia-integration

Could you please guide me/provide solutions for the following queries?

1. Am I correct in choosing the workflow-diagram-example (https://github.com/eclipse-glsp/glsp-server), or should I refer to some other example?
2. Are the activities done till now, in-line with how it should have been actually done?
3. What should be the next step in this process? What changes need to be done, and at what places, in the server component, so that the server could cater to the new diagram model?
4. Currently, the diagram that would be drawn on the Eclipse Theia client is serialized as JSON when viewed via the Code Editor. What changes needs to be done in order for the diagram to be serialized as **XML**?

Please let me know if you need any more details.

**Comment by vhemery** [comment, marked as answer]
Workflow diagram looks like a good example to me.
You can also take a look at the Coffee editor ( https://github.com/eclipse-emfcloud/coffee-editor/ ) if you want to see other features such as json forms. The workflow diagram is also integrated in it.

In any case, starting by generating the domain plugin like you did is the mandatory first step. The "model" or "src-gen" are the classic way to structure it. If your sources were generated in an "src" directory instead, you can configure this in the .genmodel file, but it is only a matter of configuration and should not be blocking for future steps.

For the next steps, you must configure and launch your model server component, before connecting a client (probably based on Theia) to it. For these steps, the Coffee editor would provide you a better understanding, as it contains "examples" branches which will make it easier to understand the process step by step.

Finally, the XML or JSON format is just a matter of serialization and encoding. This should be configurable within the Guice modules. But it's kind of the last step and easy to implement once you've mastered the rest of the framework.

**Reply by IngaleAbhishek** [reply]
Hi @vhemery, Thank you for your suggestions. I was able to create a working sample for my use case, by using the workflow-diagram example as a reference point.

Now, the only task that remains is the serialization of the diagram in XML format. I couldn't get a grasp of what needs to be done for changing the serialization format from JSON to XML in the **workflow digram** example.
Any help on this (where/what changes need to be done) would be greatly appreciated.

Thank you.

**Reply by vhemery** [reply]
I honestly haven't tried it (and thought it would be based on EMF and easier), but...

In your Guice module, you should be able to override the `bindSourceModelLoader` method. This should let you load diagrams from another format by defining a different binding.

For the persistence, the current implementation should be
https://github.com/eclipse-glsp/glsp-server/blob/415c0cb5e49efe35035dd132a99cd7850c263e75/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/actions/SaveModelActionHandler.java#L63
You should be able to override the `configureActionHandlers` method in your Guice module. You can call super, then remove the SaveModelActionHandler.class binding to replace it with your own implementation.

Hope that will help.

**Reply by IngaleAbhishek** [reply]
Thank you @vhemery, I will try out your recommendations.

**Comment by rsoika** [comment]
When you start from scratch with just a .ecore file and a .genmode file. How did you generate the code?
And how did you figure out which classes from the example workflow server are needed? For example `WorkflowServerLauncher.java` seems to be important. But all the other classes? For what are the used? Is thee a kind of "_Beginning from scratch with just a .ecore model file..._" Tutorial?

**Reply by vhemery** [reply]
> Is thee a kind of "_Beginning from scratch with just a .ecore model file..._" Tutorial?

Alas there is not to my knowledge.

The GLSP server is high-level technology which is based on other lower-level technologies.
You can start the examples by following the READMEs, but that will not give you knowledge of these underlying technologies which is a pre-requisite.

To be able to work with EMF.cloud technologies, it is important to understand first at least:
- how Dependency Injection with Guice works: https://github.com/google/guice/wiki/Motivation
- how to use and generate models with EMF: https://www.vogella.com/tutorials/EclipseEMF/article.html

Of course, if you want more than tutorials and a more professional support, you can also ask for support or training to companies providing expertise in the domain.
If you want expertise on EMF.cloud technologies, EclipseSource is probably the best candidate, since they are the main maintainers ( https://eclipsesource.com/services/training/ ).
But if you are looking for just an EMF training or other underlying technologies, you can also look at other companies among the Eclipse members ( https://www.eclipse.org/membership/exploreMembership.php ), such as OPCoach, Typefox, Itemis, Obeo...

**Note:** I am not affiliated to EclipseSource nor work for any of the companies quoted above. I just happen to like their work.

**Reply by rsoika** [reply]
hm....
I invested some more hours but I do not got out any usable starting code. Of course I am able to run the workflow example. But to isolate the workflow example into a separate code repo seems to be impossible. Looks like you need to copy all the 700(!) files of he server source code 

Yes maybe you are right and I need more training (after 40 years).

I really was impressed at the beginning but for the moment I do no longer understand what emfcloud is good for.

**Comment by rsoika** [comment]
I have now isolated the glsp workflow example in a separate project: 

https://github.com/rsoika/glsp-workflow-example

I think this makes it easier to start from scratch as it only contains the necessary source code artefacts. You can fork the project as a template to get started with your own.

**Reply by planger** [reply]
Please note that there is currently a PR in progress that provides the workflow example as a standalone: https://github.com/eclipse-glsp/glsp-examples/pull/86
Besides, we are working on providing the minimal example (also in the linked glsp-examples repo) adapted for the different technology combinations (Theia, VSCode, EMF or no EMF, with or without the Model server, etc.) to provide a simpler entry point.

Anyway, please always feel free to open issues with PRs against any of the GLSP repositories to help out making it easier to start with GLSP. Any help is greatly appreciated!

**Reply by ndoschek** [reply]
Also an update to the latest version of of the minimal-example is ready and will soon be available on master: https://github.com/eclipse-glsp/glsp-examples/pull/87

**Comment by rsoika** [comment]
In the mean time I also figured out how to define your own .ecore file based on the GLSP Definitions. Find my documentation [here](https://github.com/rsoika/glsp-workflow-example/tree/master/glsp-workflow-example-server#bring-your-own-model-extending-glsp-using-an-external-emf-model)

@IngaleAbhishek : how have you managed to get started from scratch?

**Reply by paulalexwilson** [reply]
I am also trying to understand emfcloud and specifically the model server component. I wanted to create a simple model but everything I've seen for model creation involves installing eclipse and using it to do code generation. Is this the only way?

**Reply by eneufeld** [reply]
Hi @paulalexwilson ,
yes the modelserver uses emf. So if you manage to create your emf model and generate the code without eclipse you won't need it (I would be interested in the solution ;-) ).
You can also look into the modelserver-node component which is written in typescript and currently acts as a facade in front of the modelserver so that the business logic can be completely written without java. Here you could extend the code and replace the communication with the java modelserver.

**Reply by rsoika** [reply]
@paulalexwilson I have written a short tutorial for myself. Maybe this is helpful for you:
https://github.com/imixs/open-bpmn/blob/master/doc/BPMN_EMF.md#build-the-open-bpmn-emf-model
